Enable Compound Bounding Box support

WilliamJamieson opened the pull request:
Enables support for the refactored bounding box and compound bounding box features in PRs [astropy/astropy#11930](https://github.com/astropy/astropy/pull/11930), [astropy/astropy#11931](https://github.com/astropy/astropy/pull/11931), and [astropy/astropy#11942](https://github.com/astropy/astropy/pull/11942).

This should be merged only after the astropy PRs have been merged and astropy 5.0 is released.

The CI job with dev dependencies should pass before the astropy release is out.

nden reviewed (commented):
I left one comment. 
Needs a rebase and squash and can be merged.

nden commented:
And needs a change log.

WilliamJamieson commented on gwcs/tests/test_wcs.py:
```diff
@@ -283,6 +283,60 @@ def test_bounding_box():
+        w.attach_compound_bounding_box(cbb, [('x',)])
+        from astropy.modeling.bounding_box import CompoundBoundingBox
+        cbb = CompoundBoundingBox.validate(trans3, cbb, selector_args=[('x',)], order='F')
+        assert w.bounding_box == cbb
```
Done in 6988475

WilliamJamieson commented on gwcs/tests/test_wcs.py:
```diff
@@ -283,6 +283,60 @@ def test_bounding_box():
+
+        from astropy.modeling.bounding_box import CompoundBoundingBox
+        cbb = CompoundBoundingBox.validate(trans, cbb, selector_args=[('x1',)], order='F')
+        assert w.bounding_box == cbb
```
Done in 6988475

WilliamJamieson commented:
@nden, I have added the requested `assert` statements to the tests, updated the changelog, rebased, and squashed. Please let me know if there anything else I need to do.

nden reviewed (approved):
Thanks!
